src/solvers/refinement/string_refinement_util.cpp

```diff
@@ -0,0 +1,414 @@
+/*******************************************************************\
+
+ Module: String solver
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <algorithm>
+#include <numeric>
+#include <functional>
+#include <iostream>
+#include <util/arith_tools.h>
+#include <util/ssa_expr.h>
+#include <util/std_expr.h>
+#include <util/expr_iterator.h>
+#include <util/graph.h>
+#include "string_refinement_util.h"
+
+bool is_char_type(const typet &type)
+{
+  return type.id() == ID_unsignedbv &&
+         to_unsignedbv_type(type).get_width() <= 16;
+}
+
+bool is_char_array_type(const typet &type, const namespacet &ns)
+{
+  if(type.id() == ID_symbol)
+    return is_char_array_type(ns.follow(type), ns);
+  return type.id() == ID_array && is_char_type(type.subtype());
+}
+
+bool is_char_pointer_type(const typet &type)
+{
+  return type.id() == ID_pointer && is_char_type(type.subtype());
+}
+
+bool has_subtype(
+  const typet &type,
+  const std::function<bool(const typet &)> &pred)
+{
+  if(pred(type))
+    return true;
+
+  if(type.id() == ID_struct || type.id() == ID_union)
+  {
+    const struct_union_typet &struct_type = to_struct_union_type(type);
+    return std::any_of(
+      struct_type.components().begin(),
+      struct_type.components().end(), // NOLINTNEXTLINE
+      [&](const struct_union_typet::componentt &comp) {
+        return has_subtype(comp.type(), pred);
+      });
+  }
+
+  return std::any_of( // NOLINTNEXTLINE
+    type.subtypes().begin(), type.subtypes().end(), [&](const typet &t) {
+      return has_subtype(t, pred);
+    });
+}
+
+bool has_char_pointer_subtype(const typet &type)
+{
+  return has_subtype(type, is_char_pointer_type);
+}
+
+bool has_string_subtype(const typet &type)
+{
+  // NOLINTNEXTLINE
+  return has_subtype(
+    type, [](const typet &subtype) { return subtype == string_typet(); });
+}
+
+bool has_char_array_subexpr(const exprt &expr, const namespacet &ns)
+{
+  for(auto it = expr.depth_begin(); it != expr.depth_end(); ++it)
```

Comment:
I'd personally prefer braces around the multi-line `for` block.

Reply:
I'm replacing it by a `find_if` which is more appropriate